samples/ab_testing/plugin.cc

```diff
@@ -0,0 +1,46 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START serviceextensions_ab_testing]
+#include <ctime>
+#include <string>
+
+#include "proxy_wasm_intrinsics.h"
+
+class MyHttpContext : public Context {
+ public:
+  explicit MyHttpContext(uint32_t id, RootContext* root) : Context(id, root) {}
+
+  FilterHeadersStatus onRequestHeaders(uint32_t headers,
+                                       bool end_of_stream) override {
+    std::string path = getRequestHeader(":path")->toString();
```

Comment:
toString() forces a string copy, whereas WasmDataPtr->view() will return std::string_view to avoid copies.

You may also want an existence check on the :path header, although to be fair it should always exist.

Reply:
Great point, I have added a check for the path (just in case) and changed this line to use string view.